File: app/src/main/java/org/tvheadend/tvhclient/ui/features/programs/ProgramDetailsFragment.java
package org.tvheadend.tvhclient.ui.features.programs;

import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ScrollView;
import android.widget.TextView;

import org.tvheadend.tvhclient.R;
import org.tvheadend.tvhclient.databinding.ProgramDetailsFragmentBinding;
import org.tvheadend.tvhclient.domain.entity.Program;
import org.tvheadend.tvhclient.domain.entity.Recording;
import org.tvheadend.tvhclient.ui.base.BaseFragment;
import org.tvheadend.tvhclient.util.menu.PopupMenuUtil;
import org.tvheadend.tvhclient.util.menu.SearchMenuUtils;
import org.tvheadend.tvhclient.ui.features.dvr.RecordingAddEditActivity;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.Toolbar;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.ViewModelProviders;
import timber.log.Timber;

public class ProgramDetailsFragment extends BaseFragment {

    private Toolbar nestedToolbar;
    private ScrollView scrollView;
    private TextView statusTextView;

    private int eventId;
    private int channelId;
    private Program program;
    private Recording recording;
    private int programIdToBeEditedWhenBeingRecorded = 0;
    private ProgramDetailsFragmentBinding itemBinding;

    public static ProgramDetailsFragment newInstance(int eventId, int channelId) {
        ProgramDetailsFragment f = new ProgramDetailsFragment();
        Bundle args = new Bundle();
        args.putInt("eventId", eventId);
        args.putInt("channelId", channelId);
        f.setArguments(args);
        return f;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        itemBinding = DataBindingUtil.inflate(inflater, R.layout.program_details_fragment, container, false);
        View view = itemBinding.getRoot();

        nestedToolbar = view.findViewById(R.id.nested_toolbar);
        scrollView = view.findViewById(R.id.scrollview);
        statusTextView = view.findViewById(R.id.status);
        return view;
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        forceSingleScreenLayout();

        if (!isDualPane) {
            toolbarInterface.setTitle(getString(R.string.details));
            toolbarInterface.setSubtitle("");
        }
        if (savedInstanceState != null) {
            eventId = savedInstanceState.getInt("eventId", 0);
            channelId = savedInstanceState.getInt("channelId", 0);
        } else {
            Bundle bundle = getArguments();
            if (bundle != null) {
                eventId = bundle.getInt("eventId", 0);
                channelId = bundle.getInt("channelId", 0);
            }
        }

        ProgramViewModel viewModel = ViewModelProviders.of(activity).get(ProgramViewModel.class);
        program = viewModel.getProgramByIdSync(eventId);
        if (program != null) {
            Timber.d("Loaded details for program " + program.getTitle());
            itemBinding.setProgram(program);
            itemBinding.setHtspVersion(htspVersion);
            itemBinding.setIsProgramArtworkEnabled(isUnlocked && sharedPreferences.getBoolean("program_artwork_enabled", false));
            // The toolbar is hidden as a default to prevent pressing any icons if no recording
            // has been loaded yet. The toolbar is shown here because a recording was loaded
            nestedToolbar.setVisibility(View.VISIBLE);
            activity.invalidateOptionsMenu();
        } else {
            scrollView.setVisibility(View.GONE);
            statusTextView.setText(getString(R.string.error_loading_program_details));
            statusTextView.setVisibility(View.VISIBLE);
        }

        viewModel.getRecordingsByChannelId(channelId).observe(getViewLifecycleOwner(), recordings -> {
            Timber.d("Got recordings");
            if (recordings != null) {
                boolean recordingExists = false;
                for (Recording rec : recordings) {
                    // Show the edit recording screen of the scheduled recording
                    // in case the user has selected the record and edit menu item.
                    // Otherwise remember the recording so that the state can be updated
                    if (rec.getEventId() == programIdToBeEditedWhenBeingRecorded
                            && programIdToBeEditedWhenBeingRecorded > 0) {
                        programIdToBeEditedWhenBeingRecorded = 0;
                        Intent intent = new Intent(activity, RecordingAddEditActivity.class);
                        intent.putExtra("id", rec.getId());
                        intent.putExtra("type", "recording");
                        activity.startActivity(intent);
                        break;

                    } else if (program != null
                            && rec.getEventId() == program.getEventId()) {
                        Timber.d("Found recording for program " + program.getTitle());
                        recording = rec;
                        recordingExists = true;
                        break;
                    }
                }
                // If there is no recording for the program set the
                // recording to null so that the correct state is shown
                if (!recordingExists) {
                    recording = null;
                }
                // Update the state of the recording (if there is one)
                // and also the menu items in the nested toolbar
                program.setRecording(recording);
                itemBinding.setProgram(program);
                activity.invalidateOptionsMenu();
            }
        });
    }

    @Override
    public void onPrepareOptionsMenu(@NonNull Menu menu) {
        // Show or hide search menu items in the main toolbar
        PopupMenuUtil.prepareSearchMenu(menu, program.getTitle(), isNetworkAvailable);
        // Show or hide menus of the nested toolbar
        PopupMenuUtil.prepareMenu(activity, nestedToolbar.getMenu(), program, program.getRecording(), isNetworkAvailable, htspVersion, isUnlocked);
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putInt("eventId", eventId);
        outState.putInt("channelId", channelId);
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.external_search_options_menu, menu);
        nestedToolbar.inflateMenu(R.menu.program_popup_and_toolbar_menu);
        nestedToolbar.setOnMenuItemClickListener(this::onOptionsItemSelected);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // The program might be null in case the viewmodel
        // has not yet loaded the program for the given id
        if (program == null) {
            return super.onOptionsItemSelected(item);
        }
        if (SearchMenuUtils.onMenuSelected(activity, item.getItemId(), program.getTitle(), program.getChannelId())) {
            return true;
        }
        switch (item.getItemId()) {
            case R.id.menu_record_stop:
                return menuUtils.handleMenuStopRecordingSelection(recording, null);

            case R.id.menu_record_cancel:
                return menuUtils.handleMenuCancelRecordingSelection(recording, null);

            case R.id.menu_record_remove:
                return menuUtils.handleMenuRemoveRecordingSelection(recording, null);

            case R.id.menu_record_once:
                return menuUtils.handleMenuRecordSelection(program.getEventId());

            case R.id.menu_record_once_and_edit:
                programIdToBeEditedWhenBeingRecorded = program.getEventId();
                return menuUtils.handleMenuRecordSelection(program.getEventId());

            case R.id.menu_record_once_custom_profile:
                return menuUtils.handleMenuCustomRecordSelection(program.getEventId(), program.getChannelId());

            case R.id.menu_record_series:
                return menuUtils.handleMenuSeriesRecordSelection(program.getTitle());

            case R.id.menu_play:
                return menuUtils.handleMenuPlayChannel(program.getChannelId());

            case R.id.menu_cast:
                return menuUtils.handleMenuCast("channelId", program.getChannelId());

            case R.id.menu_add_notification:
                return menuUtils.handleMenuAddNotificationSelection(program);

            default:
                return super.onOptionsItemSelected(item);
        }
    }
}

    

File: app/src/main/java/org/tvheadend/tvhclient/ui/features/programs/ProgramDetailsFragment.kt
package org.tvheadend.tvhclient.ui.features.programs

import android.os.Bundle
import android.view.*
import androidx.lifecycle.ViewModelProvider
import org.tvheadend.data.entity.Program
import org.tvheadend.data.entity.Recording
import org.tvheadend.tvhclient.R
import org.tvheadend.tvhclient.databinding.ProgramDetailsFragmentBinding
import org.tvheadend.tvhclient.ui.base.BaseFragment
import org.tvheadend.tvhclient.ui.common.*
import org.tvheadend.tvhclient.ui.common.interfaces.ClearSearchResultsOrPopBackStackInterface
import org.tvheadend.tvhclient.ui.common.interfaces.LayoutControlInterface
import org.tvheadend.tvhclient.util.extensions.gone
import org.tvheadend.tvhclient.util.extensions.visible
import timber.log.Timber

class ProgramDetailsFragment : BaseFragment(), ClearSearchResultsOrPopBackStackInterface {

    private lateinit var programViewModel: ProgramViewModel
    private var program: Program? = null
    private var recording: Recording? = null
    private var programIdToBeEditedWhenBeingRecorded = 0
    private lateinit var binding: ProgramDetailsFragmentBinding

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        //binding = DataBindingUtil.inflate(inflater, R.layout.program_details_fragment, container, false)
        binding = ProgramDetailsFragmentBinding.inflate(inflater, container, false)
        binding.layout.viewTreeObserver.addOnPreDrawListener(object : ViewTreeObserver.OnPreDrawListener {
            override fun onPreDraw(): Boolean {
                binding.layout.viewTreeObserver.removeOnPreDrawListener(this)
                binding.viewWidth = binding.layout.measuredWidth
                Timber.d("Width of program details layout is %s", binding.layout.measuredWidth)
                return true
            }
        })
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        programViewModel = ViewModelProvider(requireActivity())[ProgramViewModel::class.java]

        if (activity is LayoutControlInterface) {
            (activity as LayoutControlInterface).forceSingleScreenLayout()
        }

        toolbarInterface.setTitle(getString(R.string.details))
        toolbarInterface.setSubtitle("")

        // In case the fragment was called from the program activity, program list or any adapter
        arguments?.let {
            programViewModel.eventIdLiveData.value = it.getInt("eventId", 0)
            programViewModel.channelIdLiveData.value = it.getInt("channelId", 0)
        }

        Timber.d("Observing program")
        programViewModel.program.observe(viewLifecycleOwner) {
            Timber.d("View model returned a program")
            program = it
            showProgramDetails()
        }

        Timber.d("Observing recordings")
        programViewModel.recordings.observe(viewLifecycleOwner) { recordings ->
            if (recordings != null) {
                Timber.d("View model returned ${recordings.size} recordings")
                showRecordingStatusOfProgram(recordings)
            }
        }
    }

    private fun showProgramDetails() {
        if (program != null) {
            binding.program = program
            binding.viewModel = programViewModel
            // The toolbar is hidden as a default to prevent pressing any icons if no recording
            // has been loaded yet. The toolbar is shown here because a recording was loaded
            binding.nestedToolbar.visible()
            activity?.invalidateOptionsMenu()
        } else {
            binding.scrollview.gone()
            binding.status.text = getString(R.string.error_loading_program_details)
            binding.status.visible()
        }
    }

    private fun showRecordingStatusOfProgram(recordings: List<Recording>) {
        var recordingExists = false
        for (rec in recordings) {
            // Show the edit recording screen of the scheduled recording
            // in case the user has selected the record and edit menu item.
            // Otherwise remember the recording so that the state can be updated
            if (rec.eventId == programIdToBeEditedWhenBeingRecorded && programIdToBeEditedWhenBeingRecorded > 0) {
                programIdToBeEditedWhenBeingRecorded = 0
                editSelectedRecording(requireActivity(), rec.id)
                break

            } else {
                val p = program
                if (p != null && rec.eventId == p.eventId) {
                    Timber.d("Found recording for program ${p.title}")
                    recording = rec
                    recordingExists = true
                    break
                }
            }
        }
        // If there is no recording for the program set the
        // recording to null so that the correct state is shown
        if (!recordingExists) {
            recording = null
        }
        // Update the state of the recording (if there is one)
        // and also the menu items in the nested toolbar
        program?.recording = recording
        binding.program = program
        activity?.invalidateOptionsMenu()
    }

    override fun onPrepareOptionsMenu(menu: Menu) {
        val ctx = context ?: return

        // Hide the play button in the main toolbar, its already available in the nested toolbar
        menu.findItem(R.id.menu_play)?.isVisible = false

        preparePopupOrToolbarSearchMenu(menu, program?.title, isConnectionToServerAvailable)
        preparePopupOrToolbarRecordingMenu(ctx, binding.nestedToolbar.menu, program?.recording, isConnectionToServerAvailable, htspVersion, isUnlocked)
        preparePopupOrToolbarMiscMenu(ctx, binding.nestedToolbar.menu, program, isConnectionToServerAvailable, isUnlocked)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.external_search_options_menu, menu)
        binding.nestedToolbar.inflateMenu(R.menu.program_popup_and_toolbar_menu)
        binding.nestedToolbar.setOnMenuItemClickListener { this.onOptionsItemSelected(it) }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // The program might be null in case the view model
        // has not yet loaded the program for the given id
        val program = this.program ?: return super.onOptionsItemSelected(item)
        val ctx = context ?: return super.onOptionsItemSelected(item)
        when (item.itemId) {
            R.id.menu_stop_recording -> return showConfirmationToStopSelectedRecording(ctx, recording, null)
            R.id.menu_cancel_recording -> return showConfirmationToCancelSelectedRecording(ctx, recording, null)
            R.id.menu_remove_recording -> return showConfirmationToRemoveSelectedRecording(ctx, recording, null)
            R.id.menu_record_program -> return recordSelectedProgram(ctx, program.eventId, programViewModel.getRecordingProfile(), htspVersion)
            R.id.menu_record_program_and_edit -> {
                programIdToBeEditedWhenBeingRecorded = program.eventId
                return recordSelectedProgram(ctx, program.eventId, programViewModel.getRecordingProfile(), htspVersion)
            }
            R.id.menu_record_program_with_custom_profile -> return recordSelectedProgramWithCustomProfile(ctx, program.eventId, program.channelId, programViewModel.getRecordingProfileNames(), programViewModel.getRecordingProfile())
            R.id.menu_record_program_as_series_recording -> return recordSelectedProgramAsSeriesRecording(ctx, program.title, program.channelId, programViewModel.getRecordingProfile(), htspVersion)
            R.id.menu_play -> return playSelectedChannel(ctx, program.channelId, isUnlocked)
            R.id.menu_cast -> return castSelectedChannel(ctx, program.channelId)

            R.id.menu_search_imdb -> return searchTitleOnImdbWebsite(ctx, program.title)
            R.id.menu_search_fileaffinity -> return searchTitleOnFileAffinityWebsite(ctx, program.title)
            R.id.menu_search_youtube -> return searchTitleOnYoutube(ctx, program.title)
            R.id.menu_search_google -> return searchTitleOnGoogle(ctx, program.title)
            R.id.menu_search_epg -> return searchTitleInTheLocalDatabase(requireActivity(), baseViewModel, program.title, program.channelId)

            R.id.menu_add_notification -> return addNotificationProgramIsAboutToStart(ctx, program, programViewModel.getRecordingProfile())
            else -> return false
        }
    }

    companion object {

        fun newInstance(eventId: Int = 0, channelId: Int = 0): ProgramDetailsFragment {
            val f = ProgramDetailsFragment()
            val args = Bundle()
            args.putInt("eventId", eventId)
            args.putInt("channelId", channelId)
            f.arguments = args
            return f
        }
    }
}

    